### PR TITLE
Prevent wrong usage of Control Objects with assertions

### PR DESCRIPTION
**Pull request by Holzhaus**
This should prevent CO failures from going unnoticed. Right now, only a warning is displayed in the log, which can easily be overlooked.

Zulip discussion that prompted this PR: https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/Macros.2FSaved.20Hotcue.20Routines.2F.22Serato.20Flip.22/near/201478587

### CODE REVIEW

File: src/control/control.cpp
Comment (uklotzde) [inline src/control/control.cpp:124]: This should work:
```
VERIFY_OR_DEBUG_ASSERT(!pCreateCO) {
} else {
}
```

Also use qWarning() instead of qDebug()

File: src/control/control.cpp
Comment (uklotzde) [inline src/control/control.cpp:144]: Replace `if {...}` with `DEBUG_ASSERT(flags.testFlag(ControlFlag::NoAssertIfMissing));`

File: src/control/control.h
Comment (uklotzde) [inline src/control/control.h:18]: The negation make the code harder to understand. Why not `WarnIfMissing` and `WarnAndAssertIfMissing`? Using flags is not necessary according to the code, because the assertion is nested in the warning path. Use a simple enum class instead.

File: src/controllers/controllerengine.cpp
Comment (uklotzde) [inline src/controllers/controllerengine.cpp:1445]: This will become ControlFlag::WarnIfMissing after my suggested changes.

File: src/skin/legacyskinparser.cpp
Comment (uklotzde) [inline src/skin/legacyskinparser.cpp:105]: This will become ControlFlag::None after my suggested changes.

**Review by uklotzde — CHANGES_REQUESTED** [review]

**Comment by uklotzde** [comment]
Renamed the PR to better reflect the intention.

File: src/control/control.cpp
Comment (Holzhaus) [reply src/control/control.cpp:124]: I thought about this too, but then it instantly throws the debug assertion when `DEBUG_ASSERTIONS_FATAL` is enabled and does not print the log message before. That makes it harder to see which CO is responsible.

File: src/control/control.h
Comment (Holzhaus) [reply src/control/control.h:18]: I was planning on adding a few more flags, e.g. `DelayedInitialization` for ControlProxy objects that are created before the actual CO (e.g. `[AutoDJ],enabled`). Right now we just disable the warnings completely, but IMHO this is not a good solution.

Comment (Holzhaus) [reply src/control/control.h]: > Why not WarnIfMissing and WarnAndAssertIfMissing?

I wanted to make it explicit from the caller side that the warnings/assertions are disabled, so that don't set `InitializeLater` and accidently disable warnings/assertions because you forgot to OR this with the `WarnAndAssertIfMissing` flag.

Btw, I also had to add a flag to allow missing keys for WWidgetStack (which takes params from skins).

**Comment by Holzhaus** [comment]
I hope everything is in order now. I feel like our CO stuff could use some refactoring, e.g. the skin system uses `getControl` to let skins create COs. I also removed the confusing method to change the key of a ControlProxy after creating it through the initialize method (just use a new object if you want that) and got rid of the ControlProxy constructor without key (which is apparently not needed and also not useful anymore).

File: src/control/controlobject.h
Comment (uklotzde) [inline src/control/controlobject.h:46]: No inline required

File: src/control/controlproxy.cpp
Comment (uklotzde) [inline src/control/controlproxy.cpp:21]: But we just tested that InitializeLater is **not** set? Why then set it to false again?

File: src/control/controlproxy.cpp
Comment (uklotzde) [inline src/control/controlproxy.cpp:31]: Why not:
VERIFY_OR_DEBUG_ASSERT(!m_key.isNull() || flags.testFlag(ControlFlag::AllowEmptyKey)) {
return;
}

The extra logical or in release mode is negligible.

File: src/control/controlproxy.cpp
Comment (uklotzde) [inline src/control/controlproxy.cpp:37]: We should decide for one order. First the condition followed the the flag test or the other way round? I would prefer to use `|| testFlag()` as a suffix, because the condition is what actually matters.

File: src/control/controlproxy.h
Comment (uklotzde) [inline src/control/controlproxy.h:24]: Replace NULL with nullptr while editing legacy code

File: src/test/beatstranslatetest.cpp
Comment (uklotzde) [inline src/test/beatstranslatetest.cpp:29]: Isn't `None` the default and could be omitted here?

File: src/mixer/playermanager.cpp
Comment (daschuer) [inline src/mixer/playermanager.cpp:68]: Why is ControlFlag::NoWarnIfMissing set?

File: src/skin/legacyskinparser.cpp
Comment (daschuer) [reply src/skin/legacyskinparser.cpp:105]: Instead of Implicit disable warnings, we can introduce another function that is called getOrCreateControl() or something.

**Review by daschuer — CHANGES_REQUESTED** [review]

**Comment by daschuer** [comment]
Instead of Implicit disable warnings, we can introduce another function that is called getOrCreateControl() or something and maybe tryGetControl()

**Comment by daschuer** [comment]
Is this really a 2.3 PR? This is not without a risk.

File: src/mixer/playermanager.cpp
Comment (Holzhaus) [reply src/mixer/playermanager.cpp:68]: Oops, this should be IntializeLater:
https://github.com/mixxxdj/mixxx/pull/2896/files#diff-d5e088ea01baa5f34dd6173a0de8bf91R607

Right now this throws a warning on every start.

Comment (daschuer) [reply src/mixer/playermanager.cpp:68]: Is this a real bug than?

Comment (daschuer) [reply src/mixer/playermanager.cpp:68]: No, I have found the testing code below.

Comment (Holzhaus) [reply src/mixer/playermanager.cpp]: fixed.

**Comment by Holzhaus** [comment]
> Is this really a 2.3 PR? This is not without a risk.

I think it makes sense. It adds additional safety and the additional debug assertions will be disabled in release builds anyway.

**Review by uklotzde — APPROVED** [review]

**Comment by uklotzde** [comment]
```
Warning [Main]: ControlDoublePrivate::getControl returning NULL for ( "" , "visual_key_distance" )
Critical [Main]: DEBUG ASSERT: "flags.testFlag(ControlFlag::NoAssertIfMissing)" in function static QSharedPointer<ControlDoublePrivate> ControlDoublePrivate::getControl(const ConfigKey&, ControlFlags, ControlObject*, bool, bool, bool, double) at /tmp/mixxx/src/control/control.cpp:148
```

**Review by uklotzde — CHANGES_REQUESTED** [review]
master: Multiple debug assertions while starting Mixxx

**Comment by Holzhaus** [comment]
> master: Multiple debug assertions while starting Mixxx

You mean you pulled this branch into master?

**Comment by uklotzde** [comment]
> > master: Multiple debug assertions while starting Mixxx
> 
> You mean you pulled this branch into master?

Yes, my regular development version is based on master.

**Comment by Holzhaus** [comment]
I suppose this might be caused by the controller engine changes from #2682.

**Comment by uklotzde** [comment]
Reminder for later: The changes in `ControllerEngine` have to be fixed manually when merging from 2.3 into master!

Remaining debug assertions in master on startup:
```
Warning [Main]: ControlDoublePrivate::getControl returning NULL for ( "" , "visual_key_distance" )
Critical [Main]: DEBUG ASSERT: "flags.testFlag(ControlFlag::NoAssertIfMissing)" in function static QSharedPointer<ControlDoublePrivate> ControlDoublePrivate::getControl(const ConfigKey&, ControlFlags, ControlObject*, bool, bool, bool, double) at /tmp/mixxx/src/control/control.cpp:148
Critical [Main]: DEBUG ASSERT: "m_pControl || flags.testFlag(ControlFlag::NoAssertIfMissing)" in function void ControlProxy::initialize(ControlFlags) at /tmp/mixxx/src/control/controlproxy.cpp:35
Critical [Main]: DEBUG ASSERT: "valid() || flags.testFlag(ControlFlag::NoAssertIfMissing)" in function void ControlProxy::initialize(ControlFlags) at /tmp/mixxx/src/control/controlproxy.cpp:36
Warning [Main]: ControlDoublePrivate::getControl returning NULL for ( "" , "visual_key_distance" )
Critical [Main]: DEBUG ASSERT: "flags.testFlag(ControlFlag::NoAssertIfMissing)" in function static QSharedPointer<ControlDoublePrivate> ControlDoublePrivate::getControl(const ConfigKey&, ControlFlags, ControlObject*, bool, bool, bool, double) at /tmp/mixxx/src/control/control.cpp:148
Critical [Main]: DEBUG ASSERT: "m_pControl || flags.testFlag(ControlFlag::NoAssertIfMissing)" in function void ControlProxy::initialize(ControlFlags) at /tmp/mixxx/src/control/controlproxy.cpp:35
Critical [Main]: DEBUG ASSERT: "valid() || flags.testFlag(ControlFlag::NoAssertIfMissing)" in function void ControlProxy::initialize(ControlFlags) at /tmp/mixxx/src/control/controlproxy.cpp:36
Warning [Main]: ControlDoublePrivate::getControl returning NULL for ( "" , "visual_key_distance" )
Critical [Main]: DEBUG ASSERT: "flags.testFlag(ControlFlag::NoAssertIfMissing)" in function static QSharedPointer<ControlDoublePrivate> ControlDoublePrivate::getControl(const ConfigKey&, ControlFlags, ControlObject*, bool, bool, bool, double) at /tmp/mixxx/src/control/control.cpp:148
Critical [Main]: DEBUG ASSERT: "m_pControl || flags.testFlag(ControlFlag::NoAssertIfMissing)" in function void ControlProxy::initialize(ControlFlags) at /tmp/mixxx/src/control/controlproxy.cpp:35
Critical [Main]: DEBUG ASSERT: "valid() || flags.testFlag(ControlFlag::NoAssertIfMissing)" in function void ControlProxy::initialize(ControlFlags) at /tmp/mixxx/src/control/controlproxy.cpp:36
Warning [Main]: ControlDoublePrivate::getControl returning NULL for ( "" , "visual_key_distance" )
Critical [Main]: DEBUG ASSERT: "flags.testFlag(ControlFlag::NoAssertIfMissing)" in function static QSharedPointer<ControlDoublePrivate> ControlDoublePrivate::getControl(const ConfigKey&, ControlFlags, ControlObject*, bool, bool, bool, double) at /tmp/mixxx/src/control/control.cpp:148
Critical [Main]: DEBUG ASSERT: "m_pControl || flags.testFlag(ControlFlag::NoAssertIfMissing)" in function void ControlProxy::initialize(ControlFlags) at /tmp/mixxx/src/control/controlproxy.cpp:35
Critical [Main]: DEBUG ASSERT: "valid() || flags.testFlag(ControlFlag::NoAssertIfMissing)" in function void ControlProxy::initialize(ControlFlags) at /tmp/mixxx/src/control/controlproxy.cpp:36
```

These have to be fixed in time if we decide to merge this PR. Otherwise starting Mixxx with development settings in master becomes annoying.

I assume that these assertions do not appear in 2.3 if you tested it. No need to recompile everything twice locally. But we need to have a plan how to fix those in master.

**Review by uklotzde — APPROVED** [review]

**Comment by uklotzde** [comment]
@daschuer Merge?

**Comment by Holzhaus** [comment]
Let me recheck this first, I'll make sure that none of the assertions from master are already present here. I could be that these assertions just don't fire because skin is not lacking a CO.

File: src/control/control.h
Comment (daschuer) [inline src/control/control.h:18]: I have thought about the "InitializeLater" flag and I think we can do better. 
Let's either remove this and assert or make early initialization work. 
In any case we should remove the flag otherwise I am afraid this will be a source of a mayor confusion for new developers.  

See my other comments for fix the root issue.

File: src/mixer/playermanager.cpp
Comment (daschuer) [inline src/mixer/playermanager.cpp:67]: This can be fixed by just keeping the m_pAutoDjEnabled null.

File: src/mixer/playermanager.cpp
Comment (daschuer) [inline src/mixer/playermanager.cpp:606]: this can be replaced by 
```
if (!m_pAutoDjEnabled) {
      m_pAutoDjEnabled = make_parented(....
```
and should always successful here

**Review by daschuer — CHANGES_REQUESTED** [review]

File: src/control/control.h
Comment (daschuer) [inline src/control/control.h:18]: Do we want to start with 1?

**Comment by daschuer** [comment]
The flaky test strikes again: 

```
        Start  19: AutoDJProcessorTest.EnabledSuccess_DecksStopped_TrackLoadFails
 18/616 Test  #19: AutoDJProcessorTest.EnabledSuccess_DecksStopped_TrackLoadFails ..................***Failed    0.06 sec
qt.qpa.screen: QXcbConnection: Could not connect to display :99.0
Could not connect to any X display. 
```
can't we call mixxx-test instead of ctest?

**Comment by daschuer** [comment]
The rest LGTM, Thank you.

**Review by daschuer — APPROVED** [review]

**Comment by Holzhaus** [comment]
The `NoWarnIfMissing` flag is only used for the legacy skin parser that uses `getControl` to check if a CO already exists. We could drop it by adding a `ControlObject::exists(key)` or `ControlObject::getOrCreateControl(key)` function. Should I do that? If so, what do you prefer?

**Comment by daschuer** [comment]
Yes ControlObject::getOrCreateControl(key) would be much better express the intent.
I would prefer this, because it will require only one hash lookup.

**Comment by daschuer** [comment]
Oh sorry, scratch that.

I am fine with the current solution. It is performs good and is easy to understand.

**Comment by Holzhaus** [comment]
Ok, then it's ready to merge now.

**Comment by daschuer** [comment]
Thank you for taking care if this.

**Comment by uklotzde** [comment]
After merging 2.3 into master many tests trigger debug assertions:

Example:
```
ControlDoublePrivate::getControl returning NULL for ( "[Master]" , "num_decks" )
DEBUG ASSERT: "flags.testFlag(ControlFlag::NoAssertIfMissing)" in function static QSharedPointer<ControlDoublePrivate> ControlDoublePrivate::getControl(const ConfigKey&, ControlFlags, ControlObject*, bool, bool, bool, double) at /tmp/mixxx/src/control/control.cpp:144
DEBUG ASSERT: "m_pControl || flags.testFlag(ControlFlag::NoAssertIfMissing)" in function void ControlProxy::initialize(ControlFlags) at /tmp/mixxx/src/control/controlproxy.cpp:35
DEBUG ASSERT: "valid() || flags.testFlag(ControlFlag::NoAssertIfMissing)" in function void ControlProxy::initialize(ControlFlags) at /tmp/mixxx/src/control/controlproxy.cpp:36
a
```

**Comment by Holzhaus** [comment]
See #2910.

**Pull request merged**
